### PR TITLE
Sophos XG Firewall

### PR DESCRIPTION
**Pull request by ShiratG**
## Contributing to Cortex XSOAR Content
Make sure to register your contribution by filling the [contribution registration form](https://forms.gle/XDfxU4E61ZwEESSMA)

**The Pull Request will be reviewed only after the contribution registration form is filled.**

## Status
- [ ] In Progress
- [x] Ready

## Related Issues
reference: https://github.com/demisto/etc/issues/24149

## Minimum version of Demisto
- [x] 5.0.0

## Does it break backward compatibility?
   - [x] No

## Must have
- [x] Tests
- [x] Documentation

### CODE REVIEW

File: Packs/SophosXGFirewall/pack_metadata.json [hunk 1]
```diff
@@ -0,0 +1,16 @@
+{
+    "name": "Sophos XG Firewall",
+    "description": "Sophos XG Firewall is on-premise firewall by Sophos",
```
Comment (bakatzir) [inline Packs/SophosXGFirewall/pack_metadata.json:3]:
```suggestion
    "description": "Use the Sophos XG Firewa to manage your firewall, detect and respond to threats on your
network.",
```

File: Packs/SophosXGFirewall/Integrations/sophos_firewall.py [hunk 2]
```diff
@@ -0,0 +1,2020 @@
+import demistomock as demisto
+from CommonServerPython import *
+from CommonServerUserPython import *
+
+''' IMPORTS '''
+import json
+import requests
```
Comment (bakatzir) [inline Packs/SophosXGFirewall/Integrations/sophos_firewall.py:7]: not needed, imported in CSP

File: Packs/SophosXGFirewall/Integrations/sophos_firewall.py [hunk 3]
```diff
@@ -0,0 +1,2020 @@
...
+import urllib3
+from typing import Union, Optional, Callable
+
+# Disable insecure warnings
+urllib3.disable_warnings()
+
+''' CONSTANTS '''
+DATE_FORMAT = '%Y-%m-%dT%H:%M:%SZ'
+VERSION = '1702.1'
```
Comment (bakatzir) [inline Packs/SophosXGFirewall/Integrations/sophos_firewall.py:16]:
```suggestion
API_VERSION = '1702.1'
```

Comment (bakatzir) [reply Packs/SophosXGFirewall/Integrations/sophos_firewall.py:16]: should not it be configurable with an integration parameter?

File: Packs/SophosXGFirewall/Integrations/sophos_firewall.py [hunk 4]
```diff
@@ -0,0 +1,2020 @@
...
+                return item
+
+
+def main():
+    """
+        PARSE AND VALIDATE INTEGRATION PARAMS
+    """
+    params = demisto.params()
+    command = demisto.command()
+
```
Comment (bakatzir) [inline Packs/SophosXGFirewall/Integrations/sophos_firewall.py:1846]: let's also call the args here and pass them. (args=demisto.args())

File: Packs/SophosXGFirewall/Integrations/sophos_firewall.py [hunk 5]
```diff
@@ -0,0 +1,2020 @@
...
+        if status_message and 'Successful' in status_message:
+            message = 'ok'
+
+        return message
+
+    except DemistoException as e:
+        return e.message
+
+
+def request_builder(auth: tuple, method: str, data: tuple, operation: str) -> dict:
```
Comment (bakatzir) [inline Packs/SophosXGFirewall/Integrations/sophos_firewall.py:771]: this should be a static function in the client

File: Packs/SophosXGFirewall/Integrations/sophos_firewall.py [hunk 6]
```diff
@@ -0,0 +1,2020 @@
...
+            f'{method.title()}': {
+                '@operation': operation if operation else '',
+                data[0]: data[1]
+            }
+        }
+    }
+    return {'reqxml': json2xml(json.dumps(request_data))}
+
+
+def request_one_item_builder(name: str) -> dict:
```
Comment (bakatzir) [inline Packs/SophosXGFirewall/Integrations/sophos_firewall.py:799]: should also be a static function in the client

File: Packs/SophosXGFirewall/Integrations/sophos_firewall.py [hunk 7]
```diff
@@ -0,0 +1,2020 @@
...
+        return self.request(data, 'GET', 'get')
+
+    def set_request(self, data: tuple, operation: str) -> requests.Response:
+        return self.request(data, 'POST', 'set', operation)
+
+    def delete_request(self, data: tuple) -> requests.Response:
+        return self.request(data, 'POST', 'remove')
+
+    def get_item_by_name(self, endpoint_tag: str, name: str) -> requests.Response:
+        return self.get_request((endpoint_tag, request_one_item_builder(name)))
```
Comment (bakatzir) [inline Packs/SophosXGFirewall/Integrations/sophos_firewall.py:96]: try to call the `request` func from this func directly. it is a bit misleading.

File: Packs/SophosXGFirewall/Integrations/sophos_firewall.yml [hunk 8]
```diff
@@ -0,0 +1,4166 @@
...
+    execution: false
+    name: sophos-firewall-web-filter-delete
+    outputs:
+    - contextPath: SophosFirewall.WebFilterPolicy.Name
+      description: Name of the policy.
+      type: String
+    - contextPath: SophosFirewall.WebFilterPolicy.IsDeleted
+      description: Whether or not the policy was deleted.
+      type: String
+  dockerimage: demisto/python3:3.8.3.8715
```
Comment (bakatzir) [inline Packs/SophosXGFirewall/Integrations/sophos_firewall.yml:4158]:
```suggestion
  dockerimage: demisto/python3:3.8.6.14516
```

File: Packs/SophosXGFirewall/Integrations/sophos_firewall.yml [hunk 9]
```diff
@@ -0,0 +1,4166 @@
...
+    description: Delete an existing web filter policy.
+    execution: false
+    name: sophos-firewall-web-filter-delete
+    outputs:
+    - contextPath: SophosFirewall.WebFilterPolicy.Name
+      description: Name of the policy.
+      type: String
+    - contextPath: SophosFirewall.WebFilterPolicy.IsDeleted
+      description: Whether or not the policy was deleted.
+      type: String
```
Comment (bakatzir) [inline Packs/SophosXGFirewall/Integrations/sophos_firewall.yml:4157]: Please change for all the `IsDeleted` keys
```suggestion
      type: Bool
```

File: Packs/SophosXGFirewall/Integrations/sophos_firewall.yml [hunk 10]
```diff
@@ -0,0 +1,4166 @@
...
+  dockerimage: demisto/python3:3.8.3.8715
+  feed: false
+  isfetch: true
```
Comment (bakatzir) [inline Packs/SophosXGFirewall/Integrations/sophos_firewall.yml:4160]:
```suggestion
  isfetch: false
```

File: Packs/SophosXGFirewall/Integrations/sophos_firewall.yml [hunk 11]
```diff
@@ -0,0 +1,4166 @@
...
+  longRunning: false
+  longRunningPort: false
+  runonce: false
+  script: '-'
+  subtype: python3
```
Comment (bakatzir) [inline Packs/SophosXGFirewall/Integrations/sophos_firewall.yml:4165]:
```suggestion
  subtype: python3
  fromversion: 5.0.0
```

File: Packs/SophosXGFirewall/Integrations/sophos_firewall_test.py [hunk 12]
```diff
@@ -0,0 +1,1117 @@
...
+     - A response is returned from the API
+    Then:
+     - Ensure the correct item is returned from the function based on the arguemnts.
+    """
+    from sophos_firewall import check_error_on_response
+    good_result = json.loads(load_mock_response('rule_get.json'))
+    check_error_on_response(good_result)
+    bad_result = {'Status': 'No. of records Zero.'}
+    with pytest.raises(Exception):
+        check_error_on_response(bad_result)
```
Comment (bakatzir) [inline Packs/SophosXGFirewall/Integrations/sophos_firewall_test.py:1100]: best to split to 2 test cases.
the first with good_result is suffixed `_happy`, the second as is.

File: Packs/SophosXGFirewall/Playbooks/Sophos_Firewall_-_Block_IP.yml [hunk 13]
```diff
@@ -0,0 +1,819 @@
+id: b9ff938b-ec31-4110-88d3-390274746097
+version: 34
+vcShouldKeepItemLegacyProdMachine: false
```
Comment (bakatzir) [inline Packs/SophosXGFirewall/Playbooks/Sophos_Firewall_-_Block_IP.yml:3]:
```suggestion
id: Sophos Firewall - Block IP
version: -1
```

File: Packs/SophosXGFirewall/Playbooks/Sophos_Firewall_-_Block_URL.yml [hunk 14]
```diff
@@ -0,0 +1,797 @@
+id: 8463857c-c101-4aba-8f5a-27e133494ed9
+version: 27
+vcShouldKeepItemLegacyProdMachine: false
```
Comment (bakatzir) [inline Packs/SophosXGFirewall/Playbooks/Sophos_Firewall_-_Block_URL.yml:3]:
```suggestion
id: Sophos Firewall - Block URL
version: -1
```

**Review by bakatzir — COMMENTED** [review]
Very extensive and nice work. please tend to the 1st batch of comments, and let's talk about the Playbooks on Wednesday. especially regarding the use of extended context.

File: Packs/SophosXGFirewall/Integrations/sophos_firewall.py (hunk 3 above)
Comment (ShiratG) [reply Packs/SophosXGFirewall/Integrations/sophos_firewall.py:16]: No, the API versions are very different, so the endpoints aren't the same on other versions. Just wanted to make sure the version is clear and accessible in the code.
I use the variable only once in the code, and I can remove it, but I think it's more friendly the way it is. What do you think?

File: Packs/SophosXGFirewall/Integrations/sophos_firewall.py [hunk 15]
```diff
@@ -0,0 +1,2027 @@
...
+        client (Client): Sophos XG Firewall Client
+        constants (dict): CONSTANT with the endpoint tag and table headers
+        params (dict): params for the builder
+        builder (Callable): the builder to build the object
+
+    Returns:
+        CommandResults: Command results object
+    """
+    to_update = True
+    return generic_save(client, params, builder, to_update, **constants)
```
Comment (bakatzir) [inline Packs/SophosXGFirewall/Integrations/sophos_firewall.py:1719]: these two functions do not do the same action as generic_get/generic_save
they merely call them.
Have you considered removing them, and move the `to_update` argument in the signature of the function?

File: Packs/SophosXGFirewall/Integrations/sophos_firewall.py [hunk 16]
```diff
@@ -0,0 +1,2027 @@
...
+    Returns:
+        CommandResults: Command results object
+    """
+    data = builder(client, to_update, endpoint_tag, **params)
+    operation = 'update' if to_update else 'add'
+    response = client.set_request((endpoint_tag, data), operation)
+    response = json.loads(xml2json(response.text))
+
+    check_error_on_response(response)
+    return generic_get(client, params.get('name'), endpoint_tag, table_headers)
```
Comment (bakatzir) [inline Packs/SophosXGFirewall/Integrations/sophos_firewall.py:1743]: Here, I would only update the object and not return a response and call the `generic_get` from the prev func. e.g from `sophos_firewall_app_policy_add_command`.

**Review by bakatzir — COMMENTED** [review]
i think you have too many `# pylint: disable=unused-argument` around the code. please re visit them. and try to handle everything. we can also re check it together if you need help.

**Review by bakatzir — APPROVED** [review]

**Pull request merged**